File: api/internal/models/trace.go
package models

import "time"

type TraceContext struct {
	ID        string                 `json:"id"`
	Title     string                 `json:"title"`
	Latitude  float64                `json:"latitude"`
	Longitude float64                `json:"longitude"`
	Timestamp time.Time              `json:"timestamp"`
	Region    string                 `json:"region"`
	Category  string                 `json:"category"`
	Metadata  map[string]interface{} `json:"metadata,omitempty"`
}

type ClusterBounds struct {
	MinLat float64 `json:"minLat"`
	MinLng float64 `json:"minLng"`
	MaxLat float64 `json:"maxLat"`
	MaxLng float64 `json:"maxLng"`
}

type TraceCluster struct {
	ID        string        `json:"id"`
	Latitude  float64       `json:"latitude"`
	Longitude float64       `json:"longitude"`
	Count     int           `json:"count"`
	IsCluster bool          `json:"isCluster"`
	Bounds    ClusterBounds `json:"bounds"`
	Event     *TraceContext `json:"event,omitempty"`
}

type ClusteredResponse struct {
	Time        time.Time      `json:"time"`
	TotalEvents int            `json:"totalEvents"`
	Clusters    []TraceCluster `json:"clusters"`
}

type TimelineSummary struct {
	StartTime   time.Time   `json:"startTime"`
	EndTime     time.Time   `json:"endTime"`
	TimeSlices  []time.Time `json:"timeSlices"`
	TotalEvents int         `json:"totalEvents"`
}

File: web/src/types/trace.ts
export interface TraceContext {
  id: string;
  title: string;
  latitude: number;
  longitude: number;
  timestamp: string;
  region: string;
  category: string;
  metadata?: Record<string, any>;
}

export interface ClusterBounds {
  minLat: number;
  minLng: number;
  maxLat: number;
  maxLng: number;
}

export interface TraceCluster {
  id: string;
  latitude: number;
  longitude: number;
  count: number;
  isCluster: boolean;
  bounds: ClusterBounds;
  event?: TraceContext;
}

export interface ClusteredResponse {
  time: string;
  totalEvents: number;
  clusters: TraceCluster[];
}

export interface TimelineSummary {
  startTime: string;
  endTime: string;
  timeSlices: string[];
  totalEvents: number;
}
